rln/src/iden3calc/graph.rs

```diff
@@ -0,0 +1,957 @@
+                if let Some(c) = match op {
+                    Eq | Leq | Geq => Some(true),
+                    Neq | Lt | Gt => Some(false),
+                    _ => None,
+                } {
+                    nodes[i] = Node::Constant(U256::from(c));
+                    constants += 1;
+                }
+            }
+        } else if let Node::UnoOp(op, a) = nodes[i] {
+            if let Node::Constant(va) = nodes[a] {
+                nodes[i] = Node::Constant(op.eval(va));
+                constants += 1;
+            }
+        } else if let Node::TresOp(op, a, b, c) = nodes[i] {
+            if let (Node::Constant(va), Node::Constant(vb), Node::Constant(vc)) =
+                (nodes[a], nodes[b], nodes[c])
+            {
+                nodes[i] = Node::Constant(op.eval(va, vb, vc));
+                constants += 1;
+            }
+        }
+    }
+
+    eprintln!("Propagated {constants} constants");
+}
+
+/// Remove unused nodes
+pub fn tree_shake(nodes: &mut Vec<Node>, outputs: &mut [usize]) {
+    assert_valid(nodes);
+
+    // Mark all nodes that are used.
+    let mut used = vec![false; nodes.len()];
+    for &i in outputs.iter() {
+        used[i] = true;
+    }
+
+    // Work backwards from end as all references are backwards.
+    for i in (0..nodes.len()).rev() {
+        if used[i] {
+            if let Node::Op(_, a, b) = nodes[i] {
+                used[a] = true;
+                used[b] = true;
+            }
+            if let Node::UnoOp(_, a) = nodes[i] {
+                used[a] = true;
+            }
+            if let Node::TresOp(_, a, b, c) = nodes[i] {
+                used[a] = true;
+                used[b] = true;
+                used[c] = true;
+            }
+        }
+    }
+
+    // Remove unused nodes
+    let n = nodes.len();
+    let mut retain = used.iter();
+    nodes.retain(|_| *retain.next().unwrap());
+    let removed = n - nodes.len();
+
+    // Renumber references.
+    let mut renumber = vec![None; n];
+    let mut index = 0;
+    for (i, &used) in used.iter().enumerate() {
+        if used {
+            renumber[i] = Some(index);
+            index += 1;
+        }
+    }
+    assert_eq!(index, nodes.len());
+    for (&used, renumber) in used.iter().zip(renumber.iter()) {
+        assert_eq!(used, renumber.is_some());
+    }
+
+    // Renumber references.
+    for node in nodes.iter_mut() {
+        if let Node::Op(_, a, b) = node {
+            *a = renumber[*a].unwrap();
+            *b = renumber[*b].unwrap();
+        }
+        if let Node::UnoOp(_, a) = node {
+            *a = renumber[*a].unwrap();
+        }
+        if let Node::TresOp(_, a, b, c) = node {
+            *a = renumber[*a].unwrap();
+            *b = renumber[*b].unwrap();
+            *c = renumber[*c].unwrap();
+        }
+    }
+    for output in outputs.iter_mut() {
+        *output = renumber[*output].unwrap();
+    }
+
+    eprintln!("Removed {removed} unused nodes");
```

Comment:
Maybe a good idea to think about a more holistic approach to logging

Reply:
I guess there is no need in these logging parts, so I am going to remove them.